clients/go/admin/deviceflow/token_orchestrator.go

```diff
@@ -0,0 +1,236 @@
+package deviceflow
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"golang.org/x/net/context/ctxhttp"
+	"golang.org/x/oauth2"
+
+	"github.com/flyteorg/flyteidl/clients/go/admin/cache"
+	"github.com/flyteorg/flyteidl/clients/go/admin/oauth"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+	"github.com/flyteorg/flytestdlib/logger"
+)
+
+const (
+	cliendID       = "client_id"
+	deviceCode     = "device_code"
+	grantType      = "grant_type"
+	scope          = "scope"
+	grantTypeValue = "urn:ietf:params:oauth:grant-type:device_code"
+)
+
+const (
+	errSlowDown    = "slow_down"
+	errAuthPending = "authorization_pending"
+)
+
+// OAuthTokenOrError containing the token
+type OAuthTokenOrError struct {
+	*oauth2.Token
+	Error string `json:"error,omitempty"`
+}
+
+// TokenOrchestrator implements the main logic to initiate device authorization flow
+type TokenOrchestrator struct {
+	cfg          Config
+	clientConfig *oauth.Config
+	tokenCache   cache.TokenCache
+}
+
+// RefreshToken attempts to refresh the access token if a refresh token is provided.
+func (t TokenOrchestrator) RefreshToken(ctx context.Context, token *oauth2.Token) (*oauth2.Token, error) {
+	ts := t.clientConfig.TokenSource(ctx, token)
+	var refreshedToken *oauth2.Token
+	var err error
+	refreshedToken, err = ts.Token()
+	if err != nil {
+		logger.Warnf(ctx, "failed to refresh the token due to %v and will be doing re-auth", err)
+		return nil, err
+	}
+
+	if refreshedToken != nil {
+		logger.Debugf(ctx, "got a response from the refresh grant for old expiry %v with new expiry %v",
+			token.Expiry, refreshedToken.Expiry)
+		if refreshedToken.AccessToken != token.AccessToken {
+			if err = t.tokenCache.SaveToken(refreshedToken); err != nil {
+				logger.Errorf(ctx, "unable to save the new token due to %v", err)
+				return nil, err
+			}
+		}
+	}
+
+	return refreshedToken, nil
+}
+
+// FetchTokenFromCacheOrRefreshIt fetches the token from cache and refreshes it if it'll expire within the
+// Config.TokenRefreshGracePeriod period.
+func (t TokenOrchestrator) FetchTokenFromCacheOrRefreshIt(ctx context.Context) (token *oauth2.Token, err error) {
+	token, err = t.tokenCache.GetToken()
+	if err != nil {
+		return nil, err
+	}
+
+	if !token.Valid() {
+		return nil, fmt.Errorf("token from cache is invalid")
+	}
+
+	// If token doesn't need to be refreshed, return it.
+	if time.Now().Before(token.Expiry.Add(-t.cfg.TokenRefreshGracePeriod.Duration)) {
+		logger.Infof(ctx, "found the token in the cache")
+		return token, nil
+	}
+	token.Expiry = token.Expiry.Add(-t.cfg.TokenRefreshGracePeriod.Duration)
```

Comment:
What is this needed for?

Reply:
Currently there exists a bug in refresh token flow and even if the auth server return the refresh token , it never gets used . 

This is because , we use  reusableTokenSource https://github.com/golang/oauth2/blob/master/oauth2.go#L301
And which when figures out that it need to refresh the token but it can't since it has isValid check which compares the expiry.

And hence the token can't be refreshed until it expires which defeats the purpose of refresh token.

Hence we update the expiry with the grace period.

Added test to prove this issue

https://github.com/flyteorg/flyteidl/pull/313/files#diff-08990613781c77ceb555d931311713208a8d45ca6462eb52b87a98a452afe50d